CONFIG.MD
# CONFIG

Configuration of the WebService/API can be done through a **JavaScript**, **JSON** or **YAML** config file and/or through **Environment Variables**. Environment variables always have priority!

The config file can be located **anywhere on your local filesystem**. Point the API to the file by using the `-c` or `--configfile` argument or the `CONFIGFILE` environment variable(see [Startup commands](#startup-commands) below).

- If you don't provide a config file or ENV vars, the API will use the default settings (see [Default Settings](#default-settings) below)!
  - Note that the defaults settings don't have entries for Elasticsearch and Kibana! You have to use a config file to point at ELK servers/clusters!!
- JavaScript files need to export the settings (use `module.exports = {...}`).
- All of the keys inside the `env` object can be overritten with Environment Variables (for example: the `env.ES_HOST` key in the config file matches ENV var `ES_HOST`)
- All of the keys inside the `params` object are client-side defaults that the API will use in case the client does not send them in the POST body.

## Startup commands

Install type|Startup command (where `{path}` is the absolute path to the config file)
---|---
docker-compose|`$ CONFIGFILE={path} docker-compose up`
git clone|`$ node ./server.js --configfile {path}`
NPM global install|`$ timings --configfile {path}`
Docker (stand-alone)|`$ docker run -d -v {path}:/src/.config.js -p {VM_port}:{Host_port} mverkerk/timings:{version}`

NOTE: For non-docker installs, you may need elevated permissions to run the API on ports < 1024

## Default settings

```javascript

{
  env: {                          // Server-side settings
    ES_PROTOCOL: "http",            // Scheme used by the ElasticSearch server/cluster (default: "http")
    ES_HOST: "",                    // Hostname of the ElasticSearch server/cluster (default: "")
    ES_PORT: 9200,                  // Scheme used by the ElasticSearch server/cluster (default: 9200)
    ES_TIMEOUT: 5000,               // Timeout for ElasticSearch server/cluster (default: 5000)
    ES_VERSION: "",                 // The version of ElasticSearch (default: '5.6.2')
    ES_USER: "",                    // Username for Basic auth to ElasticSearch server/cluster (default: "")
    ES_PASS: "",                    // Password for Basic auth to ElasticSearch server/cluster (default: "")
    ES_SSL_CERT: "",                // Path to cert file for SSL auth to ElasticSearch server/cluster (default: "")
    ES_SSL_KEY: "",                 // Path to key file for SSL auth toElasticSearch server/cluster (default: "")
    KB_HOST: "",                    // Hostname of the Kibana server/cluster (default: -copy value from ES_HOST-)
    KB_PORT: 5601,                  // Scheme used by the ElasticSearch server/cluster (default: 5601)
    KB_INDEX: "",                   // Name of the main Kibana index in ElasticSearch (default: ".kibana") - used for upgrades
    HTTP_PORT: 80                   // Port of the API server (default: 80)
  },
  // API parameters
  params: {                       // Client-side parameters
    required: [                     // REQUIRED LOG parameters (have to start with 'log.')
      log.test_info,                  // Information about the test-step
      log.env_tester,                 // Environment of the test machine (i.e. `local`, `grid`, `saucelabs`, `iphone 7`, `galaxy-s6`, etc.)
      log.team,                       // The name of the test team (this is used to organize the Kibana dashboard)
      log.browser,                    // The browser used to perform the test (i.e. `Chrome`, `FireFox`, etc. - for API tests, use something like `api-curl`)
      log.env_target                  // Environment of the target application (i.e. `dev`, `test`, `prod`)
    ],
    defaults: {                     // DEFAULT parameters (will be used if they are not provided in the client's POST body)
      baseline: {                     // These settings are used to calculate the baseline
        days: 7,                        // Number of days to calculate the baseline for (default: 7)
        perc: 75,                       // Percentile to calculate (default: 75)
        padding: 1.2                    // Multiplier to calculate extra padding on top of the baseline (default: 1.2 = 20%)
      },
      flags: {                        // These booleans determine the output and other actions to be performed
        assertBaseline: true,           // Whether or not to compare against baseline (default: true)
        debug: false,                   // Request extra debug info from the API (default: false)
        esTrace: false,                 // Request elasticsearch output from API (default: false)
        esCreate: false,                // Save results to elasticsearch (default: false)
        passOnFailedAssert: false       // Pass the test, even when the performance is above the threshold (default: false)
      }
    }
  }
}
```

## Additional Environment variables

In addition to the `env` parameters in the config file, the following environment variables can be used to control certain settings:

Variable|Description|Examples|Default value
---|---|---|---
**CONFIGFILE**|Path to custom config file|`"/etc/timings/perfconfig.js"`|if omitted, API will use `<installation path>/.config.js`
**NODE_ENV**|The node environment|`"production"`, `"test"`, `"development"`|`"development"`
**LOG_LEVEL**|The desired level of logging - used for console output and app log|`"info"`, `"debug"`, `"error"`|`"info"`
**LOG_PATH**|Custom logging path|`"/var/logs/"`, `"/home/user/logs"`|`"<installation path>/logs/timings"`
**ES_UPGRADE**|Boolean to force installation of Kibana items|`"true"` or `"false"`|`"false"`

## The "required" array

In your server config file, you can use the `params.required` array to enforce certain LOG keys. Example:

```json
params: {
  required: [log.test_info,log.env_tester,log.team,log.browser,log.env_target],
  ...
}
```

In the above example, the API will expect that the client's POST body includes a `log` object that looks like this:

```javascript
{
  log: {
    test_info: "...",
    env_tester: "...",
    team: "...",
    browser: "...",
    env_target: "...",
  }
}
```

**TIP:** use clients can use dynamic values to populate fields like `browser` and `env_target` from sources like `process.env` (node/javascript) or `os.environ` (python).

If any of the required keys are not provided, the API will return an error. Example:

```javascript
{
    status: 422,
    message: "ValidationError: child 'log' fails because [child 'test_info' fails because ['test_info' is required]]"
}
```


CHANGELOG.md
<a name="1.1.5"></a>
## 1.1.5 (2018-01-10)

* 1.1.5 ([d57f8fb](https://github.com/godaddy/timings/commit/d57f8fb))
* Release 1.1.4 ([d380054](https://github.com/godaddy/timings/commit/d380054))
* Release 1.1.5 - updates and fixes ([0a7c13f](https://github.com/godaddy/timings/commit/0a7c13f))



<a name="1.1.4"></a>
## 1.1.4 (2018-01-06)

* 1.1.4 ([ffde0e8](https://github.com/godaddy/timings/commit/ffde0e8))
* Add content-type to resources & waterfall ([7538924](https://github.com/godaddy/timings/commit/7538924))
* Add waterfall filter and KB_REPLACE config setting ([d39cbde](https://github.com/godaddy/timings/commit/d39cbde))
* Config sample set default HTTP port to 80 ([e171596](https://github.com/godaddy/timings/commit/e171596))
* Update documentation & add ES reset in GET /health* endpoint ([44d628a](https://github.com/godaddy/timings/commit/44d628a))
* Updated documentation ([6ad63fa](https://github.com/godaddy/timings/commit/6ad63fa))



<a name="1.1.3"></a>
## 1.1.3 (2018-01-02)

* 1.1.3 ([5c266e8](https://github.com/godaddy/timings/commit/5c266e8))
* Multiple updates & fixes ([6de6397](https://github.com/godaddy/timings/commit/6de6397))



<a name="1.1.2"></a>
## 1.1.2 (2017-12-20)

* 1.1.2 ([a4ba12c](https://github.com/godaddy/timings/commit/a4ba12c))
* Adding image for github documentation ([11b73cf](https://github.com/godaddy/timings/commit/11b73cf))
* Fixed issue with access.log file and fixed documentation errors ([3463428](https://github.com/godaddy/timings/commit/3463428))
* Update config file handling: ([ddad86a](https://github.com/godaddy/timings/commit/ddad86a))
* Update README.md ([a7af9e3](https://github.com/godaddy/timings/commit/a7af9e3))
* Update README.md ([12e0f52](https://github.com/godaddy/timings/commit/12e0f52))



<a name="1.1.1"></a>
## 1.1.1 (2017-12-11)

* 1.1.1 ([c7f9e79](https://github.com/godaddy/timings/commit/c7f9e79))
* Fix issue with waterfall page & parameter validation ([41a25d7](https://github.com/godaddy/timings/commit/41a25d7))



<a name="1.1.0"></a>
# 1.1.0 (2017-12-09)

* 1.1.0 ([9034a38](https://github.com/godaddy/timings/commit/9034a38))
* Updated documentation and swagger docs + adding import script ([9436e38](https://github.com/godaddy/timings/commit/9436e38))



<a name="1.0.9"></a>
## 1.0.9 (2017-12-05)

* 1.0.9 ([833cac6](https://github.com/godaddy/timings/commit/833cac6))
* Bug fix - error with `trim` command when no searchUrl parameter present ([a65a7e1](https://github.com/godaddy/timings/commit/a65a7e1))
* Update dependencies ([eeb8909](https://github.com/godaddy/timings/commit/eeb8909))



<a name="1.0.8"></a>
## 1.0.8 (2017-12-05)

* 1.0.8 ([dfb59f7](https://github.com/godaddy/timings/commit/dfb59f7))
* Updated documentation and parameter validation ([d63fccc](https://github.com/godaddy/timings/commit/d63fccc))
* chore(package): update vulnerable dependencies (#2) ([96a1bbf](https://github.com/godaddy/timings/commit/96a1bbf)), closes [#2](https://github.com/godaddy/timings/issues/2)



<a name="1.0.7"></a>
## 1.0.7 (2017-12-01)

* 1.0.7 ([712fa3a](https://github.com/godaddy/timings/commit/712fa3a))
* Version 1.0.7 - updates ([557ffe7](https://github.com/godaddy/timings/commit/557ffe7))



<a name="1.0.6"></a>
## 1.0.6 (2017-11-16)

* 1.0.6 ([cd02198](https://github.com/godaddy/timings/commit/cd02198))
* Add extra POST body validation & updated ES template ([134256a](https://github.com/godaddy/timings/commit/134256a))
* Adding Basic and SSL auth for elasticsearch ([ae77a1c](https://github.com/godaddy/timings/commit/ae77a1c))



<a name="1.0.5"></a>
## 1.0.5 (2017-11-06)

* 1.0.5 ([bf1d95a](https://github.com/godaddy/timings/commit/bf1d95a))
* Change old parameter `assertRum` to `assertBaseline` ([da40871](https://github.com/godaddy/timings/commit/da40871))



<a name="1.0.4"></a>
## 1.0.4 (2017-11-03)

* 1.0.4 ([08bac4c](https://github.com/godaddy/timings/commit/08bac4c))
* Fix default HTTP port ([ba8afa1](https://github.com/godaddy/timings/commit/ba8afa1))



<a name="1.0.3"></a>
## 1.0.3 (2017-11-03)

* 1.0.3 ([101a35e](https://github.com/godaddy/timings/commit/101a35e))
* Fix links in documentation ([285399f](https://github.com/godaddy/timings/commit/285399f))
* Updates to server config ([d0fe612](https://github.com/godaddy/timings/commit/d0fe612))



<a name="1.0.2"></a>
## 1.0.2 (2017-10-31)

* 1.0.2 ([f99463a](https://github.com/godaddy/timings/commit/f99463a))
* Remove modules option ([c390bfe](https://github.com/godaddy/timings/commit/c390bfe))



<a name="1.0.1"></a>
## 1.0.1 (2017-10-30)

* [fix] s/package/process ([3892460](https://github.com/godaddy/timings/commit/3892460))
* 1.0.1 ([29b9569](https://github.com/godaddy/timings/commit/29b9569))
* Initial ([a27de80](https://github.com/godaddy/timings/commit/a27de80))
* Initial commit ([f254519](https://github.com/godaddy/timings/commit/f254519))
* More updates to documentation ([8f05ff4](https://github.com/godaddy/timings/commit/8f05ff4))
* Small fixes and additions to main documentation ([fb5937a](https://github.com/godaddy/timings/commit/fb5937a))
* Update documentation ([814085f](https://github.com/godaddy/timings/commit/814085f))
* Update readme ([ecc6b6b](https://github.com/godaddy/timings/commit/ecc6b6b))
* Update README.md ([a5d9544](https://github.com/godaddy/timings/commit/a5d9544))
* Update README.md ([d01c0b4](https://github.com/godaddy/timings/commit/d01c0b4))
* Updated documentation ([68c6c06](https://github.com/godaddy/timings/commit/68c6c06))





src/config.js
const fs = require('fs');
const os = require('os');
const path = require('path');
const yaml = require('js-yaml');
const nconf = require('nconf');
const pkg = require('../package.json');

/* eslint no-sync: 0 */
// Add arguments and environment variables to nconf
nconf
  .argv({
    c: {
      alias: 'configfile',
      describe: 'path to config file'
    }
  })
  .env({
    lowerCase: true,
    whitelist: ['configfile', 'node_env', 'api_host', 'host', 'http_port', 'log_level', 'log_path', 'kb_index', 'kb_rename',
      'es_upgrade', 'es_host', 'es_port', 'es_protocol', 'es_timeout', 'es_user', 'es_passwd', 'es_ssl_cert', 'es_ssl_key', 'kb_host', 'kb_port'
    ],
    parseValues: true
  });

// Load defaults first - use './.config.js for docker-compose
let cfgFile = path.resolve(__dirname, '../.config.js');
let cfgConfig = require(cfgFile);

if (nconf.get('configfile') && fs.existsSync(path.resolve(nconf.get('configfile')))) {
  // User supplied config file - overwrite config!
  cfgFile = path.resolve(nconf.get('configfile'));
  const cfgFileExt = cfgFile.substr((cfgFile.lastIndexOf('.') + 1));
  if (cfgFileExt === 'js') {
    cfgConfig = require(cfgFile);
  } else {
    cfgConfig = yaml.safeLoad(fs.readFileSync(cfgFile, { encoding: 'utf-8' }));
  }
}

// Check for missing keys
if (!cfgConfig.env) {
  cfgConfig.env = {};
}
if (!cfgConfig.params) {
  cfgConfig.params = {};
}
if (!cfgConfig.params.defaults) {
  cfgConfig.params.defaults = {};
}
if (!cfgConfig.params.defaults.baseline) {
  cfgConfig.params.defaults.baseline = {};
}
if (!cfgConfig.params.defaults.flags) {
  cfgConfig.params.defaults.flags = {};
}

// Combine ENV variables with config - ENV is leading!
const cfgNconf = {
  env: {
    ES_PROTOCOL: nconf.get('es_protocol') || cfgConfig.env.ES_PROTOCOL || 'http',
    ES_HOST: nconf.get('es_host') || cfgConfig.env.ES_HOST || nconf.get('kb_host') ||  cfgConfig.env.KB_HOST || '',
    ES_PORT: nconf.get('es_port') || cfgConfig.env.ES_PORT || 9200,
    ES_TIMEOUT: nconf.get('es_timeout') || cfgConfig.env.ES_TIMEOUT || 5000,
    ES_USER: nconf.get('es_user') || cfgConfig.env.ES_USER || '',
    ES_PASS: nconf.get('es_pass') || cfgConfig.env.ES_PASS || '',
    ES_SSL_CERT: nconf.get('es_ssl_cert') || cfgConfig.env.ES_SSL_CERT || '',
    ES_SSL_KEY: nconf.get('es_ssl_key') || cfgConfig.env.ES_SSL_KEY || '',
    KB_HOST: nconf.get('kb_host') || cfgConfig.env.KB_HOST || nconf.get('es_host') || cfgConfig.env.ES_HOST || '',
    KB_PORT: nconf.get('kb_port') || cfgConfig.env.KB_PORT || 5601,
    KB_INDEX: nconf.get('kb_index') || cfgConfig.env.KB_INDEX || '.kibana',
    KB_RENAME: nconf.get('kb_rename') || cfgConfig.env.KB_RENAME || '',
    HTTP_PORT: nconf.get('http_port') || cfgConfig.env.HTTP_PORT || 80,
    APP_NAME: pkg.name,
    // APP_VERSION: pkg['timings-api'].api_version || '0.0.0',
    APP_CONFIG: cfgFile,
    HOST: nconf.get('api_host') || nconf.get('host') || os.hostname(),
    NODE_ENV: nconf.get('node_env') || 'development',
    INDEX_PERF: 'cicd-perf',
    INDEX_RES: 'cicd-resource',
    INDEX_ERR: 'cicd-errorlog',
    useES: false
  },
  params: {
    required: cfgConfig.params.required || ['log.test_info', 'log.env_tester', 'log.team', 'log.browser', 'log.env_target'],
    defaults: {
      baseline: {
        days: cfgConfig.params.defaults.baseline.days || 7,
        perc: cfgConfig.params.defaults.baseline.perc || 75,
        padding: cfgConfig.params.defaults.baseline.padding || 1.2
      },
      flags: {
        assertBaseline: cfgConfig.params.defaults.flags.assertBaseline || true,
        debug: cfgConfig.params.defaults.flags.debug || false,
        esTrace: cfgConfig.params.defaults.flags.esTrace || false,
        esCreate: cfgConfig.params.defaults.flags.esCreate || false,
        passOnFailedAssert: cfgConfig.params.defaults.flags.passOnFailedAssert || false
      }
    }
  }
};

// Finally, load config object into nconf
nconf
  .use('memory')
  .defaults(cfgNconf);
